### PR TITLE
cmd/tailscale/cli: add progress to tailscale file cp

### PR DESCRIPTION
**Pull request by twitchyliquid64**
Trying to send some large movie files and not knowing if it was working was a vibe, not the vibe

Looks kinda like this:

```shell
[nix-shell:~/tailscale]$ go build -o tailscale ./cmd/tailscale && sudo ./tailscale --socket=/tmp/ts/ts1.sock file cp ~/Downloads/redacted.mp4 tom-workhorse:
redacted.mp4                                          		59616/713225 kb 	8.36%	6944 kb/s
```

### CODE REVIEW

File: cmd/tailscale/cli/file.go [hunk 1]
```diff
@@ -144,30 +162,78 @@ func runCp(ctx context.Context, args []string) error {
...
+						fmt.Fprintln(os.Stderr)
+						return
+					case <-time.After(time.Second):
+						n := fileContents.ReadBytes()
+						contentLengthStr := "???"
+						if contentLength > 0 {
+							contentLengthStr = fmt.Sprint(contentLength / 1024)
+						}
+
+						fmt.Fprintf(os.Stderr, "\r\x1b[K%s\t\t%s", padTruncateString(name, 64), padTruncateString(fmt.Sprintf("%d/%s kb", n/1024, contentLengthStr), 16))
```
Comment (bradfitz) [inline cmd/tailscale/cli/file.go:199]: can you use some helpers so we don't have raw VT escapes in here?

File: cmd/tailscale/cli/file.go [hunk 2]
```diff
@@ -144,30 +162,78 @@ func runCp(ctx context.Context, args []string) error {
...
+			go func() {
+				defer wg.Done()
+				var lastBytesRead uint64
+				for {
+					select {
+					case <-done:
+						fmt.Fprintln(os.Stderr)
+						return
+					case <-time.After(time.Second):
+						n := fileContents.ReadBytes()
```
Comment (bradfitz) [inline cmd/tailscale/cli/file.go:193]: this is all indented pretty deep. can you move the goroutine out to its own named func?

File: cmd/tailscale/cli/file.go [hunk 3]
```diff
@@ -49,6 +52,21 @@ var fileCmd = &ffcli.Command{
 	},
 }
 
+type countingReader struct {
+	io.Reader
+	n uint64
```
Comment (bradfitz) [inline cmd/tailscale/cli/file.go:57]: atomic.Uint64, not uint64

File: cmd/tailscale/cli/file.go [hunk 4]
```diff
@@ -49,6 +52,21 @@ var fileCmd = &ffcli.Command{
...
+}
+
+func (c *countingReader) Read(buf []byte) (int, error) {
+	n, err := c.Reader.Read(buf)
+	atomic.AddUint64(&c.n, uint64(n))
+	return n, err
+}
+
+func (c *countingReader) ReadBytes() uint64 {
```
Comment (bradfitz) [inline cmd/tailscale/cli/file.go:66]: ... then this is kinda useless. caller can do `n.Load()` just as easily/clearly

File: cmd/tailscale/cli/file.go [hunk 5]
```diff
@@ -144,30 +162,78 @@ func runCp(ctx context.Context, args []string) error {
...
+		wg.Wait()
 	}
 	return nil
 }
 
+func padTruncateString(str string, truncateAt int) string {
+	if len(str) <= truncateAt {
+		return str + strings.Repeat(" ", truncateAt-len(str))
+	}
+	return str[:truncateAt-1] + "…"
```
Comment (bradfitz) [inline cmd/tailscale/cli/file.go:234]: this truncates in the middle of a utf-8 sequence

File: cmd/tailscale/cli/file.go (hunk 1 above)
Comment (twitchyliquid64) [reply cmd/tailscale/cli/file.go]: Done.

File: cmd/tailscale/cli/file.go (hunk 2 above)
Comment (twitchyliquid64) [reply cmd/tailscale/cli/file.go]: Done.

File: cmd/tailscale/cli/file.go (hunk 4 above)
Comment (twitchyliquid64) [reply cmd/tailscale/cli/file.go]: Done.

File: cmd/tailscale/cli/file.go (hunk 3 above)
Comment (twitchyliquid64) [reply cmd/tailscale/cli/file.go]: Done.

File: cmd/tailscale/cli/file.go (hunk 5 above)
Comment (twitchyliquid64) [reply cmd/tailscale/cli/file.go]: Made it respect code-point boundaries.

**Comment by twitchyliquid64** [comment]
Re-assigning due to OOO.

File: cmd/tailscale/cli/file.go [hunk 6]
```diff
@@ -144,30 +158,91 @@ func runCp(ctx context.Context, args []string) error {
...
+		}
+	}
+}
+
+func padTruncateString(str string, truncateAt int) string {
+	if len(str) <= truncateAt {
+		return str + strings.Repeat(" ", truncateAt-len(str))
+	}
+
+	// Truncate the string, but respect unicode codepoint boundaries.
```
Comment (mihaip) [inline cmd/tailscale/cli/file.go:236]: Could probably use https://pkg.go.dev/golang.org/x/exp/utf8string#String.Slice, but it may be overkill.

Comment (twitchyliquid64) [reply cmd/tailscale/cli/file.go:236]: Its not currently in the deps, so ill stick to the naive implementation for now.

**Pull request merged**